test/e2e/operator_test.go

```diff
@@ -4179,3 +4180,63 @@ u3YLAbyW/lHhOCiZu2iAI8AbmXem9lW6Tr7p/97s0w==
 
 	return secret, cl.Create(context.TODO(), secret)
 }
+
+// TestReconcileInternalService verifies that the operator reverts changes to
+// the router-internal service.
+func TestReconcileInternalService(t *testing.T) {
+	t.Parallel()
+	icName := types.NamespacedName{
+		Namespace: operatorNamespace,
+		Name:      "reconcile-internal-service",
+	}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { assertIngressControllerDeleted(t, kclient, ic) })
+
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, availableConditionsForPrivateIngressController...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	serviceName := controller.InternalIngressControllerServiceName(ic)
+	service := &corev1.Service{}
+	if err := kclient.Get(context.TODO(), serviceName, service); err != nil {
+		t.Fatal(err)
+	}
+
+	findPort := func(name string, service *corev1.Service) *corev1.ServicePort {
```

Comment:
Shouldn't this e2e test find/update the port on the IngressController, and check to make sure it is updated on the Service?

Reply:
`TestHostNetworkPortBinding` verifies that updating `IngressController.spec.endpointPublishingStrategy.hostNetwork.statsPort` updates the deployment.  `TestReconcileInternalService` verifies that the "metrics" port exists, and that the operator is managing it.  :thinking:... Would it make sense to add a check to `TestHostNetworkPortBinding` to verify that the service uses "metrics" (and not a number) as the target port?

Reply:
https://github.com/openshift/cluster-ingress-operator/pull/1052/commits/3a56b8dbad66a5ca58fdd7d140985872e119c173 adds an assertion on the expected ports in `TestHostNetworkPortBinding`.

Reply:
`TestHostNetworkPortBinding` would have prevented the original bug if it was testing the same thing.  It is testing `assertContainerHasPort(t, routerContainer, "metrics", 9936)`.  Doesn't it need to assert that the service now has the port?

Reply:
Or, maybe the new assertion for expected ports should have the new port numbers too?  https://github.com/openshift/cluster-ingress-operator/pull/1052/files/cfba4d9ad778d01a6ab874682ad0c2d95f81eab0#diff-cf4b4e5424070a666a364d1d1b04011478d888d6d3d65c943ca7783333b7a6e4R1018-R1034

Reply:
At least during the second check on https://github.com/openshift/cluster-ingress-operator/pull/1052/files/cfba4d9ad778d01a6ab874682ad0c2d95f81eab0#diff-cf4b4e5424070a666a364d1d1b04011478d888d6d3d65c943ca7783333b7a6e4R1093

Reply:
> `TestHostNetworkPortBinding` would have prevented the original bug if it was testing the same thing as the bug report. It is testing `assertContainerHasPort(t, routerContainer, "metrics", 9936)`. Doesn't it need to assert that the service now has the port?

Is the assertion that I added what you expected?

> Or, maybe the new assertion for expected ports should have the new port numbers too? https://github.com/openshift/cluster-ingress-operator/pull/1052/files/cfba4d9ad778d01a6ab874682ad0c2d95f81eab0#diff-cf4b4e5424070a666a364d1d1b04011478d888d6d3d65c943ca7783333b7a6e4R1018-R1034

The port changes on the pod, but the port doesn't need to change on the service if the service just references the target port by name instead of by number.  That's the idea behind #864.  The fault is that the update logic didn't actually work properly, and so the service would still have port 1936 instead of port "metrics".